plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java

```diff
@@ -0,0 +1,309 @@
+        verify(!reservedChannelsForUpdate.isEmpty(), "Update primary key itself is not supported");
+
+        this.pageSinkId = pageSinkId;
+        this.insertSink = createInsertSink(session, jdbcOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+        this.updateSink = createUpdateSink(session, jdbcOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+        this.deleteSink = createDeleteSink(session, jdbcMergeTableHandle.getDeleteTableHandle(), jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private List<Integer> getReservedChannelsForUpdate(JdbcOutputTableHandle outputTableHandle)
+    {
+        Set<String> excludedColumnNames = ImmutableSet.copyOf(mergeRowIdFieldNames);
+        List<String> allDataColumns = outputTableHandle.getColumnNames();
+        ImmutableList.Builder<Integer> reservedChannelsBuilder = ImmutableList.builder();
+        for (int channel = 0; channel < allDataColumns.size(); channel++) {
+            String column = allDataColumns.get(channel);
+            if (!excludedColumnNames.contains(column)) {
+                reservedChannelsBuilder.add(channel);
+            }
+        }
+
+        return reservedChannelsBuilder.build();
+    }
+
+    protected ConnectorPageSink createInsertSink(
+            ConnectorSession session,
+            JdbcOutputTableHandle jdbcOutputTableHandle,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        return new JdbcPageSink(session, jdbcOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    protected ConnectorPageSink createUpdateSink(
+            ConnectorSession session,
+            JdbcOutputTableHandle jdbcOutputTableHandle,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        ImmutableList.Builder<String> updateColumnNames = ImmutableList.builder();
+        ImmutableList.Builder<Type> updateColumnTypes = ImmutableList.builder();
+        for (int channel : reservedChannelsForUpdate) {
+            updateColumnNames.add(jdbcOutputTableHandle.getColumnNames().get(channel));
+            updateColumnTypes.add(jdbcOutputTableHandle.getColumnTypes().get(channel));
+        }
+        updateColumnNames.addAll(mergeRowIdFieldNames);
+        updateColumnTypes.addAll(mergeRowIdFieldTypes);
+
+        JdbcOutputTableHandle updateOutputTableHandle = new JdbcOutputTableHandle(
+                catalogName,
+                schemaName,
+                tableName,
+                updateColumnNames.build(),
+                updateColumnTypes.build(),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty());
+        return new UpdateSink(session, updateOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    protected ConnectorPageSink createDeleteSink(
```

Comment:
same comment about static methods called in constructor

Reply:
If make this method to static how to override it. Phoenix needs to create a different `updateSink`